src/Integration.TeamExplorer.UnitTests/ReferencesTests.cs

```diff
@@ -55,14 +49,31 @@ public void MicrosoftTeamFoundationControls_EnsureCorrectVersion()
                 .GetReferencedAssemblies()
                 .FirstOrDefault(ra => ra.Name == "Microsoft.TeamFoundation.Controls")
                 ?.Version;
-            var callingAssembly = Assembly.GetCallingAssembly().GetName().Version;
 
-            callingAssembly.Should().NotBeNull();
-            tfControlsAssemblyVersion.Should().NotBeNull();
+            AssertIsCorrectMajorVersion(tfControlsAssemblyVersion.Major);
+        }
+
+        private void AssertIsCorrectMajorVersion(int dllMajorVersion)
+        {
+            var executingAssemblyLocation = Assembly.GetExecutingAssembly().Location;
+            executingAssemblyLocation.Should().NotBeNull();
 
-            // We assume that the version of the test execution engine is matching the version
-            // of the target version of VS.
-            tfControlsAssemblyVersion.Major.Should().Be(callingAssembly.Major);
+            if (executingAssemblyLocation.Contains("\\VS2015\\"))
+            {
+                dllMajorVersion.Should().Be(14);
+            }
+            else if (executingAssemblyLocation.Contains("\\VS2017\\"))
+            {
+                dllMajorVersion.Should().Be(15);
+            }
+            else if (executingAssemblyLocation.Contains("\\VS2019\\"))
+            {
+                dllMajorVersion.Should().Be(16);
+            }
+            else
+            {
+                Assert.Fail("Expecting the path to contain one of 'VS2015', 'VS2017' or 'VS2019'.");
```

Comment:
I'd prefix the message with something like "Test setup error: " to make it clear from the logs that we didn't get as far as actually checking the product i.e. we don't know whether the product code is bad or not.

Normally I'd use an `Assert.Inconclusive` for test setup failures but that doesn't fail the build, and in this case the consequences are serious enough that we wouldn't want to ship until we'd resolved the issue.